activemodel/lib/active_model/attribute_set.rb

```diff
@@ -94,6 +94,14 @@ def map(&block)
       AttributeSet.new(new_attributes)
     end
 
+    # Modify attributes such that:
+    # - Removes keys which don't exist in target_attributes
+    # - Adds keys that only exist in target_attributes.
+    def adapt_to_target!(target_attributes)
+      attributes.slice!(*target_attributes.keys)
+      attributes.merge!(target_attributes.slice!(*keys))
```

Comment:
Should this be a destructive slice?

Reply:
And I only mean in the merge call, `target_attributes.slice!`

Reply:
Could you provide a code example? I'm not sure why using `slice` is better 🤔 .

Reply:
@intrip you're destructively modifying `target_attributes` by using slice! instead of slice

It's an unexpected mutation irrelevant to the code block is the difference/danger

Reply:
Not sure what your looking for but code example?

I mean that the following produces the exact same result without permanent modifications to external attributes

```ruby
def adapt_to_target!(target_attributes)
      attributes.slice!(*target_attributes.keys)
      attributes.merge!(target_attributes.slice(*keys))
end
```

You could argue that the bang is a warning for changes in both hashes, but unless the memory savings is worth it, idk if I'd go that direction

Reply:
Oh I see your point now: your proposal doesn't work because the return value from `slice!` is different than `slice`, `slice!` returns the removed key/value pairs which in this case are all the `target_attributes` that aren't in `*keys`. Modifying `target_attributes` shouldn't be an issue because it's just a temporary value which we [ovverride thereafter](https://github.com/rails/rails/blob/3b1f90e4959a56f34dbc91bf838ecf80fa4d916c/activerecord/lib/active_record/persistence.rb#L717)